src/home/homework-luks.c

```diff
@@ -344,6 +322,62 @@ static int luks_try_passwords(
         return -ENOKEY;
 }
 
+static int luks_get_volume_key(
+                UserRecord *h,
+                struct crypt_device *cd,
+                const PasswordCache *cache,
+                void *volume_key,
+                size_t *volume_key_size,
+                key_serial_t *ret_key_serial) {
+
+        char **list;
+        size_t vks;
+        int r;
+
+        assert(h);
+        assert(cd);
+        assert(volume_key);
+        assert(volume_key_size);
+
+        if (cache && cache->volume_key) {
+                /* Shortcut: If volume key was loaded from the keyring then just use it */
+                memcpy(volume_key, cache->volume_key, cache->volume_key_size);
```

Comment:
maybe add `assert(cache->volume_key_size <= *volume_key_size);` before you memcpy() here.

I mean, I understand the intended calling semantics right: on input you provide a buffer and its size, and the buffer is then filled and the size set to what is actually used.

If so, the assert really should be there. And maybe a comment explaining this.

Reply:
Instead of an assert I made it log an error message and return -ENOMEM, like cryptsetup would in this same situation

Maybe making homework crash would be a better approach. Let me know

Reply:
Let's continue in #31552